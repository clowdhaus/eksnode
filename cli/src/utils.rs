use anyhow::Result;
use regex::Regex;
use semver::Version;

/// Extract the semantic version from the version string provided
pub fn get_semver(ver: &str) -> Result<Version> {
  let re = Regex::new(r"v?(\d+\.\d+\.\d+)(-.*)?")?;
  return match re.captures(ver) {
    Some(cap) => match cap.get(1) {
      Some(cap) => {
        let version = Version::parse(cap.as_str()).unwrap();
        Ok(version)
      }
      None => Err(anyhow::anyhow!("Unable to parse version")),
    },
    None => Err(anyhow::anyhow!("Unable to parse version")),
  };
}

pub fn get_kubelet_version() -> Result<Version> {
  todo!();
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn it_gets_semver_bare() {
    let expected = Version::parse("1.20.4").unwrap();
    let result = get_semver("1.20.4").unwrap();
    assert_eq!(result, expected);
  }

  #[test]
  fn it_gets_semver_leading() {
    let expected = Version::parse("1.20.4").unwrap();
    let result = get_semver("v1.20.4").unwrap();
    assert_eq!(result, expected);
  }

  #[test]
  fn it_gets_semver_trailing() {
    let expected = Version::parse("1.20.4").unwrap();
    let result = get_semver("v1.20.4-this.something_else").unwrap();
    assert_eq!(result, expected);
  }
}
