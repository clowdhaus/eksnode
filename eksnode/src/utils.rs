use std::{
  fs::OpenOptions,
  io::Write,
  os::unix::fs::{self, OpenOptionsExt},
  path::Path,
};

use anyhow::{anyhow, Result};
use regex_lite::Regex;
use semver::Version;

/// Extract the semantic version from the version string provided
pub fn get_semver(ver: &str) -> Result<Version> {
  let re = Regex::new(r"v?(\d+\.\d+\.\d+)(-.*)?")?;
  return match re.captures(ver) {
    Some(cap) => match cap.get(1) {
      Some(cap) => {
        let version = Version::parse(cap.as_str()).unwrap();
        Ok(version)
      }
      None => Err(anyhow!("Unable to parse version")),
    },
    None => Err(anyhow!("Unable to parse version")),
  };
}

pub struct CmdResult {
  pub stdout: String,
  pub stderr: String,
  pub status: i32,
}

/// Execute a command and return the output (stdout)
pub fn cmd_exec(cmd: &str, args: Vec<&str>) -> Result<CmdResult> {
  let output = std::process::Command::new(cmd).args(args).output();

  match output {
    Ok(output) => Ok(CmdResult {
      stdout: String::from_utf8_lossy(&output.stdout).to_string(),
      stderr: String::from_utf8_lossy(&output.stderr).to_string(),
      status: output.status.code().unwrap_or(1),
    }),
    Err(e) => Err(anyhow!("Error executing command {cmd}: {e}")),
  }
}

/// Write a file to disk, setting the file mode and owner (gid/uid)
pub fn write_file<P: AsRef<Path>>(contents: &[u8], path: P, mode: Option<u32>, chown: bool) -> Result<()> {
  let mut file = OpenOptions::new()
    .write(true)
    .create(true)
    .mode(mode.unwrap_or(0o644))
    .open(&path)?;
  file.write_all(contents)?;

  if chown {
    fs::chown(&path, Some(0), Some(0))?
  }

  Ok(())
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn it_gets_semver_bare() {
    let expected = Version::parse("1.20.4").unwrap();
    let result = get_semver("1.20.4").unwrap();
    assert_eq!(result, expected);
  }

  #[test]
  fn it_gets_semver_leading() {
    let expected = Version::parse("1.20.4").unwrap();
    let result = get_semver("v1.20.4").unwrap();
    assert_eq!(result, expected);
  }

  #[test]
  fn it_gets_semver_trailing() {
    let expected = Version::parse("1.20.4").unwrap();
    let result = get_semver("v1.20.4-this.something_else").unwrap();
    assert_eq!(result, expected);
  }

  #[test]
  fn it_gets_kubelet_version() {
    let expected = Version::parse("1.24.13").unwrap();
    // This is the format returned from `kubelet --version`
    let result = get_semver("Kubernetes v1.24.13-eks-0a21954").unwrap();
    assert_eq!(result, expected);
  }
}
