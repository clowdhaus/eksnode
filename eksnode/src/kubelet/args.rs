use std::path::Path;

use anyhow::Result;

use crate::utils;

pub const ARGS_PATH: &str = "/etc/systemd/system/kubelet.service.d/10-kubelet-args.conf";
pub const EXTRA_ARGS_PATH: &str = "/etc/systemd/system/kubelet.service.d/30-kubelet-extra-args.conf";

#[derive(Debug, Default)]
pub struct Args {
  pub node_ip: String,
  pub pod_infra_container_image: String,
  pub hostname_override: Option<String>,
  pub cloud_provider: String,
  pub container_runtime: Option<String>,
}

impl Args {
  pub fn write<P: AsRef<Path>>(&self, path: P, chown: bool) -> Result<()> {
    let end = " \\\n";
    let mut args = format!("--v=2{end}");

    args.push_str(&format!("\t--node-ip={}{end}", self.node_ip));
    args.push_str(&format!(
      "\t--pod-infra-container-image={}{end}",
      self.pod_infra_container_image
    ));
    if let Some(hostname_override) = &self.hostname_override {
      args.push_str(&format!("\t--hostname-override={}{end}", hostname_override));
    }
    args.push_str(&format!("\t--cloud-provider={}{end}", self.cloud_provider));
    if let Some(container_runtime) = &self.container_runtime {
      args.push_str(&format!("\t--container-runtime={}{end}", container_runtime));
    }

    // To ensure file content integrity
    if path.as_ref().is_file() {
      std::fs::remove_file(&path)?;
    }

    let args = args.strip_suffix(end).unwrap();
    let content = format!("[Service]\nEnvironment='KUBELET_ARGS={args}'\n",);
    utils::write_file(content.as_bytes(), path, Some(0o644), chown)
  }
}

#[derive(Debug, Default)]
pub struct ExtraArgs {
  pub args: Option<String>,
}

impl ExtraArgs {
  pub fn new(args: Option<String>) -> Self {
    Self { args }
  }

  pub fn write<P: AsRef<Path>>(&self, path: P, chown: bool) -> Result<()> {
    let args = match self.args {
      Some(ref args) => args,
      None => "",
    };

    // To ensure file content integrity
    if path.as_ref().is_file() {
      std::fs::remove_file(&path)?;
    }

    let contents = format!("[Service]\nEnvironment='KUBELET_EXTRA_ARGS={args}'\n");
    utils::write_file(contents.as_bytes(), path, Some(0o644), chown)
  }
}

#[cfg(test)]
mod tests {
  use std::io::{Read, Seek, SeekFrom};

  use tempfile::NamedTempFile;

  use super::*;

  #[test]
  fn it_creates_args() {
    let args = Args {
      node_ip: "10.0.0.1".to_string(),
      pod_infra_container_image: "k8s.gcr.io/pause:3.1".to_string(),
      hostname_override: None,
      cloud_provider: "external".to_string(),
      container_runtime: Some("remote".to_string()),
    };

    // Write to file
    let mut file = NamedTempFile::new().unwrap();
    args.write(file.path(), false).unwrap();
    file.seek(SeekFrom::Start(0)).unwrap();

    // Read back contents written to file
    let mut buf = String::new();
    file.read_to_string(&mut buf).unwrap();
    insta::assert_debug_snapshot!(buf);
  }

  #[test]
  fn it_creates_empty_extrargs() {
    let args = ExtraArgs::new(None);

    // Write to file
    let mut file = NamedTempFile::new().unwrap();
    args.write(file.path(), false).unwrap();
    file.seek(SeekFrom::Start(0)).unwrap();

    // Read back contents written to file
    let mut buf = String::new();
    file.read_to_string(&mut buf).unwrap();
    insta::assert_debug_snapshot!(buf);
  }

  #[test]
  fn it_creates_extrargs() {
    let args = ExtraArgs::new(Some("--max-pods=true".to_string()));

    // Write to file
    let mut file = NamedTempFile::new().unwrap();
    args.write(file.path(), false).unwrap();
    file.seek(SeekFrom::Start(0)).unwrap();

    // Read back contents written to file
    let mut buf = String::new();
    file.read_to_string(&mut buf).unwrap();
    insta::assert_debug_snapshot!(buf);
  }
}
