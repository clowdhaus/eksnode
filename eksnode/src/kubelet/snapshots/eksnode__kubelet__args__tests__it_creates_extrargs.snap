---
source: eksnode/src/kubelet/args.rs
expression: buf
---
""
